texastribune/files
# Adds a "Copy ID" button to context menu

#### What's this PR do?

texastribune PR: https://github.com/texastribune/texastribune/pull/4652
tt-docker-base PR: https://github.com/texastribune/tt-docker-base/pull/587

- Updates development scripts in `package.json`
   - Previously, we had used `webpack` to compile our TypeScript source code. Webpack was removed at some point but the scripts in `package.json` were not updated. We can use the `tsc` CLI instead to compile our code. I updated the `build` and `watch` command. I should note that I only tested the `watch` command a few times because I didn't really have a use for it.
   - Makes note of these scripts in `README`.
- Updates `package-lock.json` to use [lockfile version](https://docs.npmjs.com/cli/v8/configuring-npm/package-lock-json#lockfileversion) 2.
   - This was a byproduct of me using an `npm` version after v7. Most of our Docker-ized repositories still use version 1 lockfiles. I can downgrade my `npm` version to get a version 1 lockfile if that's preferred. 
- Creates a `createCopyIdButton` function.
   - This function creates a `div` in the context menu that says "Copy ID". It has an `onclick` event listener that will write the ID of the media file to the clipboard.
   - [`navigator.clipboard.writeText`](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/writeText) has good browser support. 
   - However, note that it only works in [HTTPS or localhost](https://stackoverflow.com/a/51823007) environments. When outside of these environments, `navigator.clipboard` is undefined. 
   - The only case the clipboard is undefined is when we are developing on [local.texastribune.org](http://local.texastribune.org:8000). Apparently, [local.texastribune.org](http://local.texastribune.org:8000) is not considered a localhost environment. See the manual testing for information about how to circumvent this.
   - I haven't been able to come up with a good `document.execCommand` fallback for this.
- Changes function name `createCopyButton` to `createCopyUrlButton`.
   - Given that we have a new `createCopyIdButton` function, `createCopyButton` might've been confusing as a name.
- Update `createCopyUrlButton` to use `navigator.clipboard`
   - I have a `document.execCommand` fallback for this.
   - Note that we don't have to select anything for the `document.execCommand` fallback because in order to get to the context menu, we have selected the media file already (and this selection defaults to the correct URL).
   - The production implementation without `navigator.clipboard` works on Chrome and Firefox but does not work on Safari.
   - This `navigator.clipboard` implementation works on Chrome, Firefox, and Safari (in viable environments).

#### Why are we doing this? How does it help us?

This allows the photo team to more easily copy a photo ID.

#### How should this be manually tested?

- *In the `docs` folder*
   1. Clone this repository
   2. Run `npm install`
   3. Run `npm run doc`
   4. Spin up a local server for the `doc` directory (`python -m http.server --directory=docs`)
   5. Go to <http://127.0.0.1:8000/UserInterfaceTutorial.html>
   6. Select the `memory` folder and then right click
   7. Click on the "Copy ID" button
   8. Paste somewhere. The ID should be 3
   9. Try the same with the `local` folder or try uploading your own files.
- *In the `texastribune` app*
   1. Pull the `ttfiles-copy-id` branch
   3. On Chrome, we can go to <chrome://flags> to add <http://local.texastribune.org> to the "Insecure origins treated as secure" flag. Enable the flag too.
       - `navigator.clipboard` doesn't work on <http://local.texastribune.org> without this flag. We can't navigate to localhost instead because we can't authenticate through Auth0 on localhost. We can put this on staging if we want to test it out without having to enable this flag.
   5. Then we can follow similar steps as to testing in the docs folder.

#### How should this change be communicated to end users?

Tell the photo team

#### Are there any smells or added technical debt to note?

No

#### What are the relevant tickets?

<https://airtable.com/appyo1zuQd8f4hBVx/tbloNZu8GkM52NKFR/viwPxWENGdckjxRrP/recporrDL4PmEL3aT?blocks=hide>

#### Have you done the following, if applicable:
***(optional: add explanation between parentheses)***

* [ ] Tested manually on mobile? *(I don't think its possible to right click on an iPhone)*
* [ ] Updated the documentation/wiki? *(TODO separately from this PR)*

#### TODOs / next steps:

* [ ] Update docs (I reverted the doc update here so as to not clutter the PR)
* [ ] Bump the version
